### PR TITLE
HSEARCH-2627 Use a proper JBoss Logger to log and to throw exceptions

### PR DESCRIPTION
**Pull request by mincong-h**
Hi Yoann, here's the PR for https://hibernate.atlassian.net/browse/HSEARCH-2627

### CODE REVIEW

File: jsr352/cdi/src/main/java/org/hibernate/search/jsr352/cdi/context/jpa/impl/CDIEntityManagerFactoryRegistry.java [hunk 1]
```diff
@@ -98,15 +101,11 @@ public EntityManagerFactory getDefault() {
 				return getVetoedBeanReference( beanManager, PersistenceUnitAccessor.class ).entityManagerFactory;
 			}
 			catch (RuntimeException e) {
-				throw new SearchException( "Exception while retrieving the EntityManagerFactory using @PersistenceUnit."
-						+ " This generally happens either because the persistence wasn't configured properly"
-						+ " or because there are multiple persistence units." );
+				throw log.cannotRetrieveEntityManagerFactoryInJsr352();
```
Comment (yrodiere) [inline jsr352/cdi/src/main/java/org/hibernate/search/jsr352/cdi/context/jpa/impl/CDIEntityManagerFactoryRegistry.java:104]: Just noticed I forgot to propagate the cause; I'll add a commit for that.

File: jsr352/cdi/src/main/java/org/hibernate/search/jsr352/cdi/context/jpa/impl/CDIEntityManagerFactoryRegistry.java [hunk 2]
```diff
@@ -98,15 +101,11 @@ public EntityManagerFactory getDefault() {
...
 			}
 		}
 		else if ( entityManagerFactoryInstance.isAmbiguous() ) {
-			throw new SearchException( "Multiple entity manager factories have been registered in the CDI context."
-					+ " Please provide the bean name for the selected entity manager factory to the batch indexing job through"
-					+ " the 'entityManagerFactoryReference' parameter." );
+			throw log.ambiguousEntityManagerFactoryInJsr352();
```
Comment (yrodiere) [inline jsr352/cdi/src/main/java/org/hibernate/search/jsr352/cdi/context/jpa/impl/CDIEntityManagerFactoryRegistry.java:108]: Same as above, I'll add a commit.

File: jsr352/cdi/src/main/java/org/hibernate/search/jsr352/cdi/context/jpa/impl/CDIEntityManagerFactoryRegistry.java [hunk 3]
```diff
@@ -126,14 +125,12 @@ public EntityManagerFactory get(String scopeName, String reference) {
 			case CDI_SCOPE_NAME:
 				Instance<EntityManagerFactory> instance = entityManagerFactoryInstance.select( new NamedQualifier( reference ) );
 				if ( instance.isUnsatisfied() ) {
-					throw new SearchException( "No entity manager factory available in the CDI context with this bean name: '" + reference + "'."
-							+ " Make sure your entity manager factory is a named bean." );
+					throw log.noAvailableEntityManagerFactoryInCDI(reference);
```
Comment (yrodiere) [inline jsr352/cdi/src/main/java/org/hibernate/search/jsr352/cdi/context/jpa/impl/CDIEntityManagerFactoryRegistry.java]: Funny that checkstyle let this through; there should be spaces before and after "reference".

File: jsr352/cdi/src/main/java/org/hibernate/search/jsr352/cdi/context/jpa/impl/CDIEntityManagerFactoryRegistry.java [hunk 4]
```diff
@@ -126,14 +125,12 @@ public EntityManagerFactory get(String scopeName, String reference) {
...
 				}
 				factory = instance.get();
 				break;
 			default:
-				throw new SearchException( "Unknown entity manager factory scope: '" + scopeName + "'."
-						+ " Please use a supported scope." );
+				throw log.unknownEntityManagerFactoryScope(scopeName);
```
Comment (yrodiere) [inline jsr352/cdi/src/main/java/org/hibernate/search/jsr352/cdi/context/jpa/impl/CDIEntityManagerFactoryRegistry.java]: Same as above, there should be spaces before and after "scopeName".

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/context/jpa/impl/ActiveSessionFactoryRegistry.java [hunk 5]
```diff
@@ -109,25 +107,20 @@ public EntityManagerFactory get(String scopeName, String reference) {
...
 				factory = sessionFactoriesByName.get( reference );
 				if ( factory == null ) {
-					throw new SearchException( "No entity manager factory has been created with this name yet: '" + reference + "'."
-							+ " Make sure your entity manager factory is named"
-							+ " (for instance by setting the '" + org.hibernate.cfg.AvailableSettings.SESSION_FACTORY_NAME + "' option)"
-							+ " and that the entity manager factory has already been created and wasn't closed before"
-							+ " you launch the job." );
+					throw log.cannotFindEntityManagerFactoryByName(
+							reference,
+							org.hibernate.cfg.AvailableSettings.SESSION_FACTORY_NAME
```
Comment (yrodiere) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/context/jpa/impl/ActiveSessionFactoryRegistry.java]: No need to make this a parameter, you can embed it in the message annotation on the logger.

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+					+ " the 'entityManagerFactoryReference' parameter (you may also use the 'entityManagerFactoryScope'"
+					+ " parameter for more referencing options)."
+	)
+	SearchException tooManyActiveEntityManagerFactories();
+
+	@LogMessage(level = INFO)
+	@Message(id = JSR_352_MESSAGES_START_ID + 10,
+			value = "%1$s"
+	)
+	void analyzeIndexProgress(StringBuilder stringBuilder);
```
Comment (yrodiere) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java]: Maybe we should rather pass a string here? A bit strange to require a StringBuilder.

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+	)
+	void readingEntities();
+
+	@LogMessage(level = INFO)
+	@Message(id = JSR_352_MESSAGES_START_ID + 26,
+			value = "No more results, read ends."
+	)
+	void noMoreResults();
+
+	@LogMessage(level = DEBUG)
```
Comment (yrodiere) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java]: This should definitely be `level = TRACE`. It is going to be very verbose.

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+	@LogMessage(level = INFO)
+	@Message(id = JSR_352_MESSAGES_START_ID + 31,
+			value = "entityName: '%1$s', rowsToIndex: %2$d")
+	void rowsToIndex(String entityName, long rowsToIndex);
+
+	@LogMessage(level = INFO)
+	@Message(id = JSR_352_MESSAGES_START_ID + 32,
+			value = "Job execution (id=%1$d) has status %2$s. Thread sleeps %3$d ms..."
+	)
+	void jobExecutionInfo(long executionId, BatchStatus batchStatus, int sleepInMillis);
```
Comment (yrodiere) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java]: This is never used and should be removed.

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+	@Message(id = JSR_352_MESSAGES_START_ID + 24,
+			value = "%1$s"
+	)
+	void printBound(PartitionBound partitionBound);
+
+	@LogMessage(level = DEBUG)
+	@Message(id = JSR_352_MESSAGES_START_ID + 25,
+			value = "Reading entities..."
+	)
+	void readingEntities();
```
Comment (yrodiere) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java]: This should definitely be `level = TRACE`. It is going to be very verbose.
By the way, the method using this log only reads *one* entity, so maybe the message should change.

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+	@LogMessage(level = DEBUG)
+	@Message(id = JSR_352_MESSAGES_START_ID + 27,
+			value = "Processing entity with id: '%1$s'"
+	)
+	void processEntity(Object entityId);
+
+	@LogMessage(level = DEBUG)
+	@Message(id = JSR_352_MESSAGES_START_ID + 28,
+			value = "%1$s is closed.")
+	void isClosed(String simpleClassName);
```
Comment (yrodiere) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java]: If I understand correctly, the purpose of this message is to trace the opening/closing of the docwriter. So, why not call this "closingDocWriter" and make the message "Closing DocWriter for entity %1$s"?

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+
+	@LogMessage(level = DEBUG)
+	@Message(id = JSR_352_MESSAGES_START_ID + 29,
+			value = "%1$s is opened.")
+	void isOpened(String simpleClassName);
```
Comment (yrodiere) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java]: Same as above, I think this should be renamed "openingDocWriter" and the message should be "Opening DocWriter for entity %1$s"?

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+	@Message(id = JSR_352_MESSAGES_START_ID + 10,
+			value = "%1$s"
+	)
+	void analyzeIndexProgress(StringBuilder stringBuilder);
+
+	@LogMessage(level = DEBUG)
+	@Message(id = JSR_352_MESSAGES_START_ID + 11,
+			value = "Session is closed."
+	)
+	void sessionClosed();
```
Comment (yrodiere) [inline jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java]: I don't see the point of those "xxxClosed" methods. We expect the job to close resources correctly, so we should only log stuff when the job *cannot* close it. I think we should those "xxxClosed" methods; same for `closingEverything`.

**Review by yrodiere — CHANGES_REQUESTED** [review]
Thanks! Here are my two cents. Just a few changes and we'll be good to go. I added one commit to fix a problem that I should have addressed sooner.

File: jsr352/cdi/src/main/java/org/hibernate/search/jsr352/cdi/context/jpa/impl/CDIEntityManagerFactoryRegistry.java (hunk 1 above)
Comment (yrodiere) [reply jsr352/cdi/src/main/java/org/hibernate/search/jsr352/cdi/context/jpa/impl/CDIEntityManagerFactoryRegistry.java:104]: Commit added.

File: jsr352/cdi/src/main/java/org/hibernate/search/jsr352/cdi/context/jpa/impl/CDIEntityManagerFactoryRegistry.java (hunk 2 above)
Comment (yrodiere) [reply jsr352/cdi/src/main/java/org/hibernate/search/jsr352/cdi/context/jpa/impl/CDIEntityManagerFactoryRegistry.java:108]: Commit added.

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java (hunk 10 above)
Comment (mincong-h) [reply jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java]: Actually, I was planned to use this method for logging the method `close()` in both `EntityReader` and `LuceneDocWriter`, but I only added it to the writer... That's why the method looks so weird. The `simpleClassName` should be either `EntityReader` or `LuceneDocWriter`.

I'll rework it.

File: jsr352/cdi/src/main/java/org/hibernate/search/jsr352/cdi/context/jpa/impl/CDIEntityManagerFactoryRegistry.java (hunk 3 above)
Comment (mincong-h) [reply jsr352/cdi/src/main/java/org/hibernate/search/jsr352/cdi/context/jpa/impl/CDIEntityManagerFactoryRegistry.java]: My bad, I didn't format the code after the modification! Not sure how check-style detects the different syntax..

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/context/jpa/impl/ActiveSessionFactoryRegistry.java (hunk 5 above)
Comment (mincong-h) [reply jsr352/core/src/main/java/org/hibernate/search/jsr352/context/jpa/impl/ActiveSessionFactoryRegistry.java]: Nice point!

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java (hunk 6 above)
Comment (mincong-h) [reply jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java]: +1

File: jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java (hunk 12 above)
Comment (mincong-h) [reply jsr352/core/src/main/java/org/hibernate/search/jsr352/logging/impl/Log.java]: You're right. Previously, I didn't know how to debug in my IDE, so I print log to debug—that's why there're a lot of no-sense log here.

**Review by yrodiere — APPROVED** [review]

**Comment by yrodiere** [comment]
Nice! Shashed and merged, thanks!

**Pull request merged**